Cannot find component `ProtocolIntersection\u200b`

If one module imports `ProtocolIntersection`, and then another module attempts to import it from that module, `mypy` seems unable to lookup `ProtocolIntersection` in the imported module's namespace (maybe due to the presence of a zero-width space at the end?):
```python
Traceback (most recent call last):
  File "C:\Users\evanj\GitHub\coconut\coconut\command\mypy.py", line 69, in mypy_run
    stdout, stderr, exit_code = run(args)
  File "mypy\api.py", line 72, in run
  File "mypy\api.py", line 59, in _run
  File "mypy\api.py", line 73, in <lambda>
  File "mypy\main.py", line 95, in main
  File "mypy\main.py", line 174, in run_build
  File "mypy\build.py", line 194, in build
  File "mypy\build.py", line 277, in _build
  File "mypy\build.py", line 2930, in dispatch
  File "mypy\build.py", line 3321, in process_graph
  File "mypy\build.py", line 3402, in process_fresh_modules
  File "mypy\build.py", line 2121, in fix_cross_refs
  File "mypy\fixup.py", line 53, in fixup_module
  File "mypy\fixup.py", line 136, in visit_symbol_table
  File "mypy\fixup.py", line 73, in visit_type_info
  File "mypy\fixup.py", line 138, in visit_symbol_table
  File "mypy\nodes.py", line 774, in accept
  File "mypy\fixup.py", line 146, in visit_func_def
  File "mypy\types.py", line 1869, in accept
  File "mypy\fixup.py", line 244, in visit_callable_type
  File "mypy\types.py", line 2803, in accept
  File "mypy\fixup.py", line 338, in visit_type_type
  File "mypy\types.py", line 1357, in accept
  File "mypy\fixup.py", line 209, in visit_instance
  File "mypy\fixup.py", line 344, in lookup_fully_qualified_typeinfo
  File "mypy\lookup.py", line 49, in lookup_fully_qualified
AssertionError: Cannot find component 'ProtocolIntersection\u200b' for 'typing_protocol_intersection.types.ProtocolIntersection\u200b'
```

Also note that reproduction here is a bit tricky, since it seems to depend on the current status of `.mypy_cache`, though it always shows up for me after running `mypy` enough times. I'm on Python 3.9 and `mypy` 1.2.0. `mypy.ini` is just:
```ini
[mypy]
plugins = typing_protocol_intersection.mypy_plugin
```

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -1,10 +1,8 @@
-import subprocess
 import typing
 from pathlib import Path
 
-import pytest
-
 import mypy.api
+import pytest
 
 HERE = Path(__file__).parent
 
@@ -27,8 +25,8 @@ def strip_invisible() -> typing.Callable[[str], str]:
         """Removes all zero-width spaces from the input text and strips
         whitespaces.
 
-        The need for the former was born with an ugly hack that we use to
-        trick mypyc.
+        The need for the former was born with an ugly hack that we use
+        to trick mypyc.
         """
         zero_width_space = "\u200B"
         return string.strip().replace(zero_width_space, "")

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -1,0 +1,36 @@
+import subprocess
+import typing
+from pathlib import Path
+
+import pytest
+
+import mypy.api
+
+HERE = Path(__file__).parent
+
+
+@pytest.fixture
+def run_mypy(strip_invisible: typing.Callable[[str], str]):
+    def _run_mypy(input_file: Path, no_incremental: bool = True) -> typing.Tuple[str, str]:
+        args = [str(input_file), "--config-file", str(HERE / "test-mypy.ini")]
+        if no_incremental:
+            args.append("--no-incremental")
+        stdout, stderr, _ = mypy.api.run(args)
+        return strip_invisible(stdout), strip_invisible(stderr)
+
+    return _run_mypy
+
+
+@pytest.fixture
+def strip_invisible() -> typing.Callable[[str], str]:
+    def _strip_invisible(string: str) -> str:
+        """Removes all zero-width spaces from the input text and strips
+        whitespaces.
+
+        The need for the former was born with an ugly hack that we use to
+        trick mypyc.
+        """
+        zero_width_space = "\u200B"
+        return string.strip().replace(zero_width_space, "")
+
+    return _strip_invisible

--- a/tests/test_mypy_plugin.py
+++ b/tests/test_mypy_plugin.py
@@ -1,8 +1,6 @@
 import typing
 from pathlib import Path
 
-import mypy.api
-import mypy.version
 import pytest
 
 import typing_protocol_intersection.mypy_plugin
@@ -45,14 +43,14 @@ def get_expected_stderr(contents: str) -> str:
 
 
 @pytest.fixture()
-def testcase_file(request):
+def testcase_file(request, strip_invisible):
     path = request.param
     with open(HERE / path, encoding="utf-8") as file:
         contents = file.read()
     return _TestCase(
         HERE / path,
-        _strip(get_expected_stdout(contents)),
-        _strip(get_expected_stderr(contents)),
+        strip_invisible(get_expected_stdout(contents)),
+        strip_invisible(get_expected_stderr(contents)),
     )
 
 
@@ -96,27 +94,9 @@ def testcase_file(request):
     ],
     indirect=["testcase_file"],
 )
-def test_mypy_plugin(testcase_file: _TestCase):
-    stdout, stderr = _run_mypy(testcase_file.path)
+def test_mypy_plugin(testcase_file: _TestCase, run_mypy):
+    stdout, stderr = run_mypy(testcase_file.path)
     assert (stdout.strip(), stderr.strip()) == (testcase_file.expected_stdout, testcase_file.expected_stderr)
-
-
-def _run_mypy(input_file: Path) -> typing.Tuple[str, str]:
-    stdout, stderr, _ = mypy.api.run(
-        [str(input_file), "--config-file", str(HERE / "test-mypy.ini"), "--no-incremental"]
-    )
-    return _strip(stdout), _strip(stderr)
-
-
-def _strip(string: str) -> str:
-    """Removes all zero-width spaces from the input text and strips
-    whitespaces.
-
-    The need for the former was born with an ugly hack that we use to
-    trick mypyc.
-    """
-    zero_width_space = "\u200B"
-    return string.strip().replace(zero_width_space, "")
 
 
 @pytest.mark.parametrize(

--- a/tests/test_past_issues/test_4_mypy_cache/input.py
+++ b/tests/test_past_issues/test_4_mypy_cache/input.py
@@ -1,0 +1,36 @@
+"""
+source: https://gist.github.com/evhub/5ef01a7d33582e84c9838bdbe9405cd4
+
+Note this file won't typecheck, but it's a different issue we aim to detect is not there. Not only would the file not
+typecheck, but mypy would crash with:
+  AssertionError: Cannot find component 'ProtocolIntersection\u200b' for 'typing_protocol_intersection.types.ProtocolIntersection\u200b'
+See https://github.com/klausweiss/typing-protocol-intersection/issues/4 for details
+"""
+
+from typing import Protocol
+
+
+class X(Protocol):
+    x: str
+
+
+class Y(Protocol):
+    y: str
+
+
+class XY:
+    x: str = "asdf"
+    y: str = "dfaf"
+
+
+class container:
+    from typing_protocol_intersection import ProtocolIntersection as _ProtocolIntersection
+
+    ProtocolIntersection = _ProtocolIntersection
+
+
+def foo(xy: container.ProtocolIntersection[X, Y]) -> None:
+    print(xy.x, xy.y)
+
+
+foo(XY())

--- a/tests/test_past_issues/test_4_mypy_cache/input.py
+++ b/tests/test_past_issues/test_4_mypy_cache/input.py
@@ -1,3 +1,4 @@
+# pylint: skip-file
 """
 source: https://gist.github.com/evhub/5ef01a7d33582e84c9838bdbe9405cd4
 

--- a/tests/test_past_issues/test_4_mypy_cache/test_4_mypy_cache.py
+++ b/tests/test_past_issues/test_4_mypy_cache/test_4_mypy_cache.py
@@ -1,0 +1,17 @@
+"""
+ref: https://github.com/klausweiss/typing-protocol-intersection/issues/4
+"""
+from pathlib import Path
+
+HERE = Path(__file__).parent
+
+
+def test_4_mypy_cache(run_mypy):
+    # given
+    input_file = HERE / "input.py"
+    # when
+    # This may not be 100% accurate, but when the bug was present, running mypy twice was (consistently) enough
+    # to reproduce the issue. If there's no exception after 4 runs, we assume the bug's not there.
+    for _ in range(4):
+        run_mypy(input_file, no_incremental=False)
+    # then no error

--- a/typing_protocol_intersection/mypy_plugin.py
+++ b/typing_protocol_intersection/mypy_plugin.py
@@ -23,9 +23,7 @@ SignatureContext = typing.Union[mypy.plugin.FunctionSigContext, mypy.plugin.Meth
 class ProtocolIntersectionPlugin(mypy.plugin.Plugin):
     # pylint: disable=unused-argument
 
-    def report_config_data(
-        self, ctx: mypy.plugin.ReportConfigContext
-    ) -> int:
+    def report_config_data(self, ctx: mypy.plugin.ReportConfigContext) -> int:
         # Whatever this method returns is used by mypy to determine whether a module should be checked again or if a
         # cache-loaded info will do. If the obtained value is different from the previous one, cache is invalidated.
         #

--- a/typing_protocol_intersection/mypy_plugin.py
+++ b/typing_protocol_intersection/mypy_plugin.py
@@ -25,16 +25,14 @@ class ProtocolIntersectionPlugin(mypy.plugin.Plugin):
 
     def report_config_data(
         self, ctx: mypy.plugin.ReportConfigContext
-    ) -> bool:
-        # When is_check is False, the value returned by this function will be
-        # serialized as json and saved to the mypy cache.  When is_check is
-        # True, the value returned will be compared against whatever was loaded
-        # from the cache.  This is so that a plugin can potentially invalidate
-        # the cache for a module, which is exactly what we want to do if we
-        # generated any types on a previous run; otherwise mypy will try to look
-        # them up in the cache (and subsequently crash when it fails to find
-        # them).
-        return UniqueFullname.instance_counter == 0
+    ) -> int:
+        # Whatever this method returns is used by mypy to determine whether a module should be checked again or if a
+        # cache-loaded info will do. If the obtained value is different from the previous one, cache is invalidated.
+        #
+        # As far as this plugin is concerned, this method is only used to aid generation of `UniqueFullname`s. If it
+        # wasn't for this method, mypy couldn't identify where a runtime-generated class name comes from, as it looks up
+        # the class names in the cache (and subsequently crashes when it fails to find them).
+        return UniqueFullname.instance_counter
 
     def get_type_analyze_hook(
         self, fullname: str

--- a/typing_protocol_intersection/mypy_plugin.py
+++ b/typing_protocol_intersection/mypy_plugin.py
@@ -23,6 +23,19 @@ SignatureContext = typing.Union[mypy.plugin.FunctionSigContext, mypy.plugin.Meth
 class ProtocolIntersectionPlugin(mypy.plugin.Plugin):
     # pylint: disable=unused-argument
 
+    def report_config_data(
+        self, ctx: mypy.plugin.ReportConfigContext
+    ) -> bool:
+        # When is_check is False, the value returned by this function will be
+        # serialized as json and saved to the mypy cache.  When is_check is
+        # True, the value returned will be compared against whatever was loaded
+        # from the cache.  This is so that a plugin can potentially invalidate
+        # the cache for a module, which is exactly what we want to do if we
+        # generated any types on a previous run; otherwise mypy will try to look
+        # them up in the cache (and subsequently crash when it fails to find
+        # them).
+        return UniqueFullname.instance_counter == 0
+
     def get_type_analyze_hook(
         self, fullname: str
     ) -> Optional[Callable[[mypy.plugin.AnalyzeTypeContext], mypy.types.Type]]:

